### PR TITLE
Use new [Interceptable] ctor for source gen

### LINKED ISSUE #101079: ConfigurationBindingGenerator should move to GetInterceptableLocation API
Roslyn just introduced a [new API](https://github.com/dotnet/roslyn/issues/72133) for creating `[InterceptsLocation]` attributes in generated code. This uses a new opaque encoding for the location.

We'd like our partners to onboard to this new API as soon as is feasible. (We *just* merged it so may be a day or two before arcade will start handing out compilers which contain it.)

The old path-line-column approach is going to be deprecated and support removed prior to stable release of the feature.

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/101079

This changes the use of the experimental `[Interceptable]` attribute to use the newer version provided by Roslyn.

This affects the test baselines, and depending on the environment there may be an older Roslyn present, so both the original and the new baselines are in this PR. See the linked issue as well as the `ReadMe.md` file in the changed files here to get more information.

To avoid breaking v8, this PR will eventually need to be ported to v8 since the original version of the attribute will eventually no longer be supported. It may be possible to only have the newer baseline for v9, however both are included here, and the plan is to essentially port this PR as-is to v8. Once the original version is no longer necessary, we can remove the "version 0" baselines and perhaps the infrastructure for having two baselines, assuming at that time Rosyln has no plans to add a third version.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs [hunk 1]
```diff
@@ -62,6 +66,79 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
...
+
+            if (!s_hasInitializedInterceptorVersion)
+            {
+                InterceptorVersion = DetermineInterceptableVersion();
+                s_hasInitializedInterceptorVersion = true;
+            }
+        }
+
+        internal static int s_interceptorVersion;
+        internal static int InterceptorVersion
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs:78]: >InterceptorVersion [](http://example.com/codeflow?start=28&length=18)

nit: why this is not a simple { get; set;}? I see the assert but I am not sure if worth it.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs [hunk 2]
```diff
@@ -62,6 +66,79 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
...
+        internal static MethodInfo? InterceptableLocationDataGetter { get; private set; }
+        internal static MethodInfo? InterceptableLocationVersionGetter { get; private set; }
+
+        internal static int DetermineInterceptableVersion()
+        {
+            MethodInfo? getInterceptableLocationMethod = null;
+            int? interceptableVersion = null;
+
+#if UPDATE_BASELINES
+            const string envVarName = "InterceptableAttributeVersion";
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs:104]: >InterceptableAttributeVersion [](http://example.com/codeflow?start=39&length=29)

Is this specific only to the config binder source gen? or we expect usage in other places?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs [hunk 3]
```diff
@@ -62,6 +66,79 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
...
+
+#pragma warning disable RS1035 // Do not use APIs banned for analyzers
+            string? interceptableVersionString = Environment.GetEnvironmentVariable(envVarName);
+#pragma warning restore RS1035
+            if (interceptableVersionString is not null)
+            {
+                if (int.TryParse(interceptableVersionString, out int version))
+                {
+                    interceptableVersion = version;
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs:113]: > interceptableVersion = version; [](http://example.com/codeflow?start=19&length=32)

what happen if someone set the value to any random number?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Emitter.cs [hunk 4]
```diff
@@ -88,6 +91,26 @@ public InterceptsLocationAttribute(string filePath, int line, int column)
...
+                    namespace System.Runtime.CompilerServices
+                    {
+                        using System;
+                        using System.CodeDom.Compiler;
+
+                        {{Expression.GeneratedCodeAnnotation}}
+                        [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+                        file sealed class InterceptsLocationAttribute : Attribute
+                        {
+                            public InterceptsLocationAttribute(int version, string data)
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Emitter.cs:107]: >(int version, string data [](http://example.com/codeflow?start=62&length=25)

nit: looks this is the only different part when version is 1. would it better to store it in a string template with {} and pass the filling value according to InterceptorVersion?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj [hunk 5]
```diff
@@ -7,6 +7,7 @@
     <AnalyzerLanguage>cs</AnalyzerLanguage>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <DefineConstants Condition="'$(LaunchDebugger)' == 'true'">$(DefineConstants);LAUNCH_DEBUGGER</DefineConstants>
+    <DefineConstants Condition="'$(UpdateBaselines)' == 'true'">$(DefineConstants);UPDATE_BASELINES</DefineConstants>
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj:10]: ><DefineConstants Condition="'$(UpdateBaselines)' == 'true'">$(DefineConstants);UPDATE_BASELINES</DefineConstants> [](http://example.com/codeflow?start=4&length=113)

Love it :-)

**Review by tarekgh — APPROVED** [review]
Left minor comments, LGTM otherwise. 

We need to decide if we continue to go with the environment variable. Will be good if @RikkiGibson look too. Thanks!

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Emitter.cs (hunk 4 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Emitter.cs]: Sure thanks

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs]: Changed - at one point it had additional logic.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs (hunk 2 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs]: Just here. I removed the const and just inlined the string.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.cs:99]: I did add a check to throw an exception in a newer commit.

**Pull request merged**